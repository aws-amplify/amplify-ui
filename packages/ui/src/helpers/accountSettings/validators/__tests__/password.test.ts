import { Amplify } from 'aws-amplify';

import {
  getDefaultPasswordValidators,
  getMinLengthValidator,
  hasLowerCase,
  hasNumber,
  hasSpecialChar,
  hasUpperCase,
} from '../password';

const amplifyPasswordConfig = {
  aws_cognito_password_protection_settings: {
    passwordPolicyMinLength: 8,
    passwordPolicyCharacters: [
      'REQUIRES_LOWERCASE',
      'REQUIRES_NUMBERS',
      'REQUIRES_SYMBOLS',
      'REQUIRES_UPPERCASE',
    ],
  },
};

const configureSpy = jest.spyOn(Amplify, 'configure');

describe('getMinLengthValidator', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('returns minLength validator as expected', () => {
    const validator = getMinLengthValidator(4);
    expect(validator).toMatchObject({
      handler: expect.any(Function),
      message: 'Password must have at least 4 characters',
      validationMode: 'onTouched',
    });
  });

  it('validates to true when password is long enough', () => {
    const { handler } = getMinLengthValidator(4);
    const isValid = handler('longpassword');
    expect(isValid).toBe(true);
  });

  it('validates to false when password is too short', () => {
    const { handler } = getMinLengthValidator(4);
    const isValid = handler('hi');
    expect(isValid).toBe(false);
  });
});

describe('hasLowerCase', () => {
  it('validates to true when password has lower case', () => {
    const { handler } = hasLowerCase;
    const isValid = handler('password');
    expect(isValid).toBe(true);
  });

  it('validates to false when password has no lower case', () => {
    const { handler } = hasLowerCase;
    const isValid = handler('PASSWORD');
    expect(isValid).toBe(false);
  });
});

describe('hasUpperCase', () => {
  it('validates to true when password has upper case', () => {
    const { handler } = hasUpperCase;
    const isValid = handler('PASSWORD');
    expect(isValid).toBe(true);
  });

  it('validates to false when password has no upper case', () => {
    const { handler } = hasUpperCase;
    const isValid = handler('password');
    expect(isValid).toBe(false);
  });
});

describe('hasNumber', () => {
  it('validates to true when password has number', () => {
    const { handler } = hasNumber;
    const isValid = handler('password123');
    expect(isValid).toBe(true);
  });

  it('validates to false when password has no number', () => {
    const { handler } = hasNumber;
    const isValid = handler('password');
    expect(isValid).toBe(false);
  });
});

describe('hasSpecialChar', () => {
  it('validates to true when password has special character', () => {
    const { handler } = hasSpecialChar;
    const isValid = handler('password!');
    expect(isValid).toBe(true);
  });

  it('validates to false when password has no special character', () => {
    const { handler } = hasSpecialChar;
    const isValid = handler('password');
    expect(isValid).toBe(false);
  });
});

describe('getDefaultPasswordValidators', () => {
  it('returns empty array if passwordSettings is empty', () => {
    configureSpy.mockReturnValue({});
    const validators = getDefaultPasswordValidators();
    expect(validators).toStrictEqual([]);
  });

  it.only('returns validators as expected for nonempty passwordSetting', () => {
    configureSpy.mockReturnValue(amplifyPasswordConfig);
    const validators = getDefaultPasswordValidators();
    expect(validators).toMatchSnapshot();
  });
});
