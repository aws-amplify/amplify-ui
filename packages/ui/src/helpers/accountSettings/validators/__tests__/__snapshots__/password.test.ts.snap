// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDefaultPasswordValidators returns validators as expected for nonempty passwordSetting 1`] = `
Array [
  Object {
    "handler": [Function],
    "message": "Password must have at least 8 characters",
    "validationMode": "onTouched",
  },
  Object {
    "handler": [Function],
    "message": "Password must have lower case letters",
    "validationMode": "onTouched",
  },
  Object {
    "handler": [Function],
    "message": "Password must have numbers",
    "validationMode": "onTouched",
  },
  Object {
    "handler": [Function],
    "message": "Password must have special characters",
    "validationMode": "onTouched",
  },
  Object {
    "handler": [Function],
    "message": "Password must have upper case letters",
    "validationMode": "onTouched",
  },
]
`;
